### PR TITLE
[Pre Commit] infrastructure enhancements

### PR DESCRIPTION
**Pull request by ilaner**
This adds the following features:
1. Support env property inside the docker hook, to provide extra env variables to the hook.
2. support copy_files inside the docker hook, to support copy files to the integration script folders dynamically (the conftest.py has to be in the same directory of running tests) 
3. Get the test image without pulling if it exists, and pull in the background (if it's not a dry run)
3. Filters `ruff` files by py files only, significantly speeds up execution.
4. Run validate -a if pre-commit -a, because the CLI can't manage so many inputs to validate and it crashes.
5. Delete all run-unit-tests command.
6. Add support for running specific hooks.
7. Run docker hooks in multiprocessing.
8. Skipped hooks will not be generated in the config file.
9. Force colorful output
10. Add support for system hooks. system hooks will run in the same interpreter that pre-commit runs and not in virtual environment (useful for `demisto-sdk commands`)
11. Moved the command to `typer`.

fixes: [CIAC-8862](https://jira-dc.paloaltonetworks.com/browse/CIAC-8862)
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-8987

needs content small PR:
related: [CIAC-7992](https://jira-dc.paloaltonetworks.com/browse/CIAC-7992)
related:  [CIAC-7993](https://jira-dc.paloaltonetworks.com/browse/CIAC-7993)

related conent PR: https://github.com/demisto/content/pull/31224

### CODE REVIEW

File: demisto_sdk/__main__.py [hunk 1]
```diff
@@ -3493,12 +3493,6 @@ def update_content_graph(
     "--mode",
     help="Special mode to run the pre-commit with",
 )
-@click.option(
-    "-ut/--no-ut",
-    "--unit-test/--no-unit-test",
-    help="Whether to run unit tests for content items",
-    default=False,
-)
```
Comment (GuyAfik) [inline demisto_sdk/__main__.py:3501]: breaking change, please make sure the state that in changelog

File: demisto_sdk/commands/pre_commit/hooks/docker.py [hunk 2]
```diff
@@ -202,27 +218,31 @@ def prepare_hook(
         logger.debug(
             f"Elapsed time to gather tags to files: {end_time - start_time} seconds"
         )
+        if copy_files := self._get_property("copy_files"):
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/hooks/docker.py:221]: dont understand what we're doing here.
Copying something from root content path to the directory of each integration? Why?

File: demisto_sdk/commands/pre_commit/tests/test_data/.pre-commit-config_template-test.yaml [hunk 3]
```diff
@@ -78,6 +90,10 @@ repos:
       - id: should_be_skipped # only for the unit test
         skip: true
 
+      - id: hook_needs_docker
+        needs:
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/tests/test_data/.pre-commit-config_template-test.yaml:94]: whats this?

File: demisto_sdk/commands/pre_commit/hooks/hook.py [hunk 4]
```diff
@@ -1,10 +1,13 @@
+import os
 import re
 from copy import deepcopy
 from pathlib import Path
 from typing import Any, Dict, Iterable, List, Set
 
 from demisto_sdk.commands.common.logger import logger
 
+PROPERTIES_TO_DELETE = {"needs"}
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/hooks/hook.py:9]: whats this?

**Review by JudahSchwartz — COMMENTED** [review]
looks good! couple questions and comments

File: demisto_sdk/commands/pre_commit/tests/test_data/.pre-commit-config_template-test.yaml (hunk 3 above)
Comment (ilaner) [reply demisto_sdk/commands/pre_commit/tests/test_data/.pre-commit-config_template-test.yaml:94]: a list of hooks which needed by the hook.
In our use-case, we need `pytest-in-docker` to finish before running `coverage-analyze`

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,9 @@
 * Fixed an issue where **validate** command failed on valid complex layout rules and triggers.
 * Fixed an issue where *validate* command failed with release notes files for assets modeling rules folder.
 * Added support for `skip` property in **pre-commit** hooks.
+* Added support for `env`, `copy_files` property in **pre-commit** docker hooks.
```
[the remark this answers is not in the document]
Comment (ilaner) [reply CHANGELOG.md:8]: documented in pre commit readme, please review

File: demisto_sdk/commands/content_graph/objects/integration_script.py [hunk 6]
```diff
@@ -50,6 +50,12 @@ def docker_images(self) -> List[str]:
     def is_powershell(self) -> bool:
         return self.type == "powershell"
 
+    @property
+    def additional_test_requirements(self) -> Optional[List[str]]:
```
[the remark this answers is not in the document]
Comment (ilaner) [reply demisto_sdk/commands/content_graph/objects/integration_script.py:54]: forgot to delete, deleting

File: demisto_sdk/commands/pre_commit/hooks/docker.py [hunk 7]
```diff
@@ -101,40 +103,57 @@ def docker_tag_to_runfiles(
...
+        base_image=image_tag,
+        container_type=TYPE_PWSH if is_powershell else TYPE_PYTHON,
+        push=docker_login(docker_client=init_global_docker_client()),
+        should_pull=False,
+        log_prompt="DockerHook",
+    )
+    if not errors:
+        if not dry_run:
+            # pull images in background
+            subprocess.Popen(
```
[the remark this answers is not in the document]
Comment (ilaner) [reply demisto_sdk/commands/pre_commit/hooks/docker.py:132]: right

File: demisto_sdk/commands/pre_commit/hooks/docker.py (hunk 2 above)
Comment (ilaner) [reply demisto_sdk/commands/pre_commit/hooks/docker.py:221]: sometimes we need some files to be inside of the directory.
For example, `conftest.py` cannot be provided as an argument and in powershell until we fix PSModulePath we will need to copy demistomock.ps1 and commonserver

File: demisto_sdk/commands/pre_commit/hooks/hook.py (hunk 4 above)
Comment (ilaner) [reply demisto_sdk/commands/pre_commit/hooks/hook.py:9]: Makes sure that added properties (which are not supported by pre-commit) deleted in the hook.

File: demisto_sdk/commands/common/docker_helper.py [hunk 8]
```diff
@@ -344,9 +371,16 @@ def pull_or_create_test_image(
...
             "\n".join(sorted(pip_requirements)).encode("utf-8")
         ).hexdigest()
+
         test_docker_image = (
             f'{base_image.replace("demisto", "devtestdemisto")}-{identifier}'
         )
+        if not should_pull and self.get_image_from_registry(test_docker_image):
+            return test_docker_image, errors
+
+        if os.getenv("CONTENT_GITLAB_CI") and "code.pan.run" not in test_docker_image:
```
[the remark this answers is not in the document]
Comment (ilaner) [reply demisto_sdk/commands/common/docker_helper.py:381]: it is the docker helper. prefer to extract to function?

**Comment by ilaner** [comment]
@JudahSchwartz 
PR is ready for rereview.
linked relevant issues.

File: demisto_sdk/commands/common/docker_helper.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (JudahSchwartz) [reply demisto_sdk/commands/common/docker_helper.py]: yes

File: demisto_sdk/commands/pre_commit/README.md [hunk 9]
```diff
@@ -33,6 +33,27 @@ You can set this as follows.
...
+
+### Skip key
+In order to skip certain hook from running, you can add a `skip` key to the hook configuration.
+```yaml
+- id: sample-hook
+  skip: true
+```
+This key could be use together with [mode](#modes), to skip certain hook when running in a specific mode.
+
+# Needs key
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/README.md:47]: how are you deciding the header level here 🤨

File: demisto_sdk/commands/pre_commit/README.md [hunk 10]
```diff
@@ -33,6 +33,27 @@ You can set this as follows.
...
+Needs keys allow to define dependencies between hooks. If a hook with `needs` is skipped, hooks that depend on it will also be skipped.
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/README.md:48]: needs doesnt guarantee order? are you sure we dont want to just make them use skip:true for this? Seems like we're complicating things

**Review by JudahSchwartz — APPROVED** [review]
Amazing work!!!!

File: demisto_sdk/commands/pre_commit/pre_commit_command.py [hunk 11]
```diff
@@ -190,61 +209,222 @@ def prepare_hooks(
...
+
+    def _hooks_need_docker(self):
+        return {
+            hook_id
+            for hook_id, hook in self.hooks.items()
+            if (needs := hook["hook"].pop("needs", None))
+            and any("in-docker" in need for need in needs)
+        }
+
+    def _get_docker_and_no_docker_hooks(
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/pre_commit_command.py:248]: docstrings to all these functions please

File: demisto_sdk/commands/pre_commit/README.md (hunk 10 above)
Comment (ilaner) [reply demisto_sdk/commands/pre_commit/README.md:48]: oops

**Pull request merged**